File: Python/VendedoresAndArticulos.py
# STRUCT PARA LOS 10 PRODUCTOS
class Producto:
    def __init__(self):
        self.Codigo = ''
        self.Cuota = 0.0
        self.CantidadVendida = 0

# STRUCT PARA LA CAPACITACION
class Capacitacion:
    def __init__(self):
        self.Curso = ''
        self.Horas = 0
        self.Institucion = ''
        self.Costo = 0.0

# STRUCT PARA EL VENDEDOR
class Vendedor:
    def __init__(self):
        self.Nombre = ''
        self.Cedula = 0
        self.Direccion = ''
        self.Correo = ''
        self.CantidadCursos = 0
        self.Producto = [Producto() for _ in range(10)]
        self.Capacitacion = [Capacitacion() for _ in range(5)]

# Funcion para pedir numeros positivos
def menores_cero(num):
    while num < 0:
        print("Número no válido. Ingrese otro: ")
        num = float(input())
    return num

# Funcion para cargar los datos
def cargar_datos(vendedor):
    print("Ingresando datos para información")
    print("Nombre del vendedor: ")
    vendedor.Nombre = input()
    print("Cedula: ")
    vendedor.Cedula = int(input())
    print("Direccion: ")
    vendedor.Direccion = input()
    print("Correo: ")
    vendedor.Correo = input()
    print("Cantidad de Cursos  (Solo de 3 a 5 Cursos): ")
    vendedor.CantidadCursos = int(input())
    while vendedor.CantidadCursos < 3 or vendedor.CantidadCursos > 5:
        print("Número no válido. Ingrese otro: ")
        vendedor.CantidadCursos = int(input())

    print("Ingresando datos para los productos")
    for i in range(10):
        print("Codigo: ")
        vendedor.Producto[i].Codigo = input()

        print("Cantidad: ")
        vendedor.Producto[i].CantidadVendida = menores_cero(int(input()))

        print("Cuota: ")
        vendedor.Producto[i].Cuota = menores_cero(float(input()))

    for j in range(5):
        print("Nombre del curso: ")
        vendedor.Capacitacion[j].Curso = input()

        print("Cantidad de horas del curso: ")
        vendedor.Capacitacion[j].Horas = menores_cero(int(input()))

        print("Nombre de la institucion: ")
        vendedor.Capacitacion[j].Institucion = input()

        print("Costo del curso en Bs: ")
        vendedor.Capacitacion[j].Costo = menores_cero(float(input()))

# Funcion para ver cuál producto no cumple con la cuota
def mostrar_producto_sin_cuota(vendedor):
    print(f"Producto sin cubrir la cuota para el vendedor {vendedor.Nombre}:")
    for i in range(10):
        if vendedor.Producto[i].CantidadVendida < vendedor.Producto[i].Cuota:
            print(f"El producto: {vendedor.Producto[i].Codigo} no cumple la cuota")

# Funcion para encontrar al vendedor con más dinero invertido en cursos
def mas_inversion_cursos(vendedores):
    inversion_total = 0.0
    mayor_inversion = 0.0
    vendedor_indice = 0

    for i in range(20):
        for j in range(vendedores[i].CantidadCursos):
            inversion_total += vendedores[i].Capacitacion[j].Costo
        if inversion_total > mayor_inversion:
            mayor_inversion = inversion_total
            vendedor_indice = i
    print(f"El vendedor con mayor inversión en cursos es {vendedores[vendedor_indice].Nombre}")
    print(f"Con una cantidad de dinero invertida de: {mayor_inversion}")

def producto_mas_vendido(vendedores):
    indice_producto = 0
    indice_vendedor = 0
    cantidad_mayor = 0

    for i in range(20):
        for j in range(10):
            if vendedores[i].Producto[j].CantidadVendida > cantidad_mayor:
                cantidad_mayor = vendedores[i].Producto[j].CantidadVendida
                indice_vendedor = i
                indice_producto = j
    print(f"El producto más vendido es: {vendedores[indice_vendedor].Producto[indice_producto].Codigo}")

# Funcion para buscar curso por nombre
def mostrar_info_curso(vendedores, opc):
    print(f"Instituciones que ofrecen el curso {opc} y su costo")
    for i in range(20):
        for j in range(vendedores[i].CantidadCursos):
            if vendedores[i].Capacitacion[j].Curso == opc:
                print(f"Vendedor: {vendedores[i].Nombre}")
                print(f"Institucion: {vendedores[i].Capacitacion[j].Institucion}")
                print(f"Costo: {vendedores[i].Capacitacion[j].Costo}")

if __name__ == "__main__":
    vendedores = [Vendedor() for _ in range(20)]
    opc = ''

    for i in range(20):
        cargar_datos(vendedores[i])

    # Llamada a las funciones creadas
    for i in range(20):
        mostrar_producto_sin_cuota(vendedores[i])

    producto_mas_vendido(vendedores)

    mas_inversion_cursos(vendedores)

    print("Buscar curso por nombre")
    while opc != '0':
        print("Escriba el nombre del curso a buscar o escriba '0' para salir")
        opc = input()
        if opc != '0':
            mostrar_info_curso(vendedores, opc)

File: JavaScript/Ejercicios Generales/VendedoresAndArticulos.js
//PRODUCTOS
class Producto {
  constructor() {
    this.Codigo = "";
    this.Cuota = 0.0;
    this.CantidadVendida = 0;
  }
}

//CAPACITACION
class Capacitacion {
  constructor() {
    this.Curso = "";
    this.Horas = 0;
    this.Institucion = "";
    this.Costo = 0.0;
  }
}

//VENDEDOR
class Vendedor {
  constructor() {
    this.Nombre = "";
    this.Cedula = 0;
    this.Direccion = "";
    this.Correo = "";
    this.CantidadCursos = 0;
    this.Producto = Array(10).fill(new Producto());
    this.Capacitacion = Array(5).fill(new Capacitacion());
  }
}

// Funcion para pedir numeros positivos
function MenoresCero(num) {
  do {
    if (num < 0) {
      console.log("Numero no valido. Ingresar otro: ");
    }
  } while (num < 0);
  return num;
}

// Funcion para cargar los datos
function CargarDatos(Vendedores) {
  console.log("Ingresando datos para informacion");
  console.log("Nombre del vendedor: ");
  Vendedores.Nombre = prompt();

  console.log("Cedula: ");
  Vendedores.Cedula = parseInt(prompt());

  console.log("Direccion: ");
  Vendedores.Direccion = prompt();

  console.log("Correo: ");
  Vendedores.Correo = prompt();

  console.log("Cantidad de Cursos  (Solo de 3 a 5 Cursos): ");
  do {
    Vendedores.CantidadCursos = parseInt(prompt());
  } while (Vendedores.CantidadCursos < 3 || Vendedores.CantidadCursos > 5);

  console.log("Ingresando datos para los productos");
  for (let i = 0; i < 10; i++) {
    console.log("Codigo: ");
    Vendedores.Producto[i].Codigo = prompt();

    console.log("Cantidad: ");
    Vendedores.Producto[i].CantidadVendida = MenoresCero(parseInt(prompt()));

    console.log("Cuota: ");
    Vendedores.Producto[i].Cuota = MenoresCero(parseFloat(prompt()));
  }

  for (let j = 0; j < 5; j++) {
    console.log("Nombre del curso: ");
    Vendedores.Capacitacion[j].Curso = prompt();

    console.log("Cantidad de horas del curso: ");
    Vendedores.Capacitacion[j].Horas = MenoresCero(parseInt(prompt()));

    console.log("Nombre de la institucion: ");
    Vendedores.Capacitacion[j].Institucion = prompt();

    console.log("Costo del curso en Bs: ");
    Vendedores.Capacitacion[j].Costo = MenoresCero(parseFloat(prompt()));
  }
}

// Funcion para ver cual producto no cumple con la cuota
function MostrarProductoSinCuota(Vendedores) {
  console.log(
    `Producto sin cubrir la cuota para el vendedor ${Vendedores.Nombre}:`
  );
  for (let i = 0; i < 10; i++) {
    if (Vendedores.Producto[i].CantidadVendida < Vendedores.Producto[i].Cuota) {
      console.log(
        `El producto: ${Vendedores.Producto[i].Codigo} no cumple la cuota`
      );
    }
  }
}

// Funcion para encontrar al vendedor con mas dinero invertido en cursos
function MasInversionCursos(Vendedores) {
  let inversionTotal = 0.0;
  let mayorInversion = 0.0;
  let vendedorIndice = 0;

  for (let i = 0; i < 20; i++) {
    for (let j = 0; j < Vendedores[i].CantidadCursos; j++) {
      inversionTotal += Vendedores[i].Capacitacion[j].Costo;
    }
    if (inversionTotal > mayorInversion) {
      mayorInversion = inversionTotal;
      vendedorIndice = i;
    }
  }
  console.log(
    `El vendedor con mayor inversion en cursos es ${Vendedores[vendedorIndice].Nombre}`
  );
  console.log(`Con una cantidad de dinero invertida de: ${mayorInversion}`);
}

// Funcion para encontrar el producto mas vendido
function ProductoMasVendido(Vendedores) {
  let IndiceProducto = 0;
  let IndiceVendedor = 0;
  let CantidadMayor = 0;

  for (let i = 0; i < 20; i++) {
    for (let j = 0; j < 10; j++) {
      if (Vendedores[i].Producto[j].CantidadVendida > CantidadMayor) {
        CantidadMayor = Vendedores[i].Producto[j].CantidadVendida;
        IndiceVendedor = i;
        IndiceProducto = j;
      }
    }
  }
  console.log(
    `El producto mas vendido es: ${Vendedores[IndiceVendedor].Producto[IndiceProducto].Codigo}`
  );
}

// Funcion para buscar curso por nombre
function MostrarInfoCurso(vendedores, opc) {
  console.log(`Instituciones que ofrecen el curso ${opc} y su costo`);
  for (let i = 0; i < 20; i++) {
    for (let j = 0; j < vendedores[i].CantidadCursos; j++) {
      if (vendedores[i].Capacitacion[j].Curso === opc) {
        console.log(`Vendedor: ${vendedores[i].Nombre}`);
        console.log(
          `Institucion: ${vendedores[i].Capacitacion[j].Institucion}`
        );
        console.log(`Costo: ${vendedores[i].Capacitacion[j].Costo}`);
      }
    }
  }
}

function main() {
  let Vendedores = Array(20).fill(new Vendedor());
  let opc;

  for (let i = 0; i < 20; i++) CargarDatos(Vendedores[i]);

  // Llamada a las funciones creadas
  for (let i = 0; i < 20; i++) MostrarProductoSinCuota(Vendedores[i]);

  ProductoMasVendido(Vendedores);

  MasInversionCursos(Vendedores);

  console.log("Buscar curso por nombre");
  do {
    console.log(
      "Escriba el nombre del curso a buscar o escriba '0' para salir"
    );
    opc = prompt();
    if (opc !== "0") {
      MostrarInfoCurso(Vendedores, opc);
    }
  } while (opc === "0");
}

main();
